### ISSUE TITLE
unix attributes

### ISSUE THREAD

**Issue by empereira** [cannatag/ldap3#905]
hello,

How I can add a new user with UNIX attributes, like that:

```
msSFU30Name
msSFU30NisDomain
uidNumber
```
uuiNumber need to be the last value of msSFU30MaxUidNumber + 1.

This I've already managed to do, what still doesn't work is the above attributes created in DC, both with conn.add() and MODIFY_ADD.

**Comment by zorn96** [comment]
Hi @empereira! could you provide a bit more clarity on what you're doing and how it's failing? maybe a code snippet that's encountering a problem?

I'm not quite such what you mean when you say "what still doesn't work is the above attributes created in DC"

if there's no stack trace/error, then if you could also print the details of the response your connection receives when the operation fails, that would be helpful, as it will be important for determining if the issue is an encoding/attribute constraints issue or something else

**Comment by empereira** [comment]
hello @zorn96, thanks for the fast answer!!! :D

This piece of code to add a new user in our samba4 AD.

```python
if not user_exists_in_ad(remove_zeros(homolog["matricula"])):
    if conn.add(
            f'CN={homolog["nome"]},{aluno_dn}', ['inetOrgPerson', 'posixAccount', 'top'],
            {
                'cn': homolog["nome"],
                'givenName': str(homolog["nome"]).split(" ", 1)[0],
                'sn': str(homolog["nome"]).split(" ", 1)[1],
                'displayName': homolog["nome"],
                'sAMAccountName': remove_zeros(homolog["matricula"]),
                'mail': f'{remove_zeros(homolog["matricula"])}@aluno.company.org',
                'userPrincipalName': f'{remove_zeros(homolog["matricula"])}@company.org',
                'userAccountControl': '512',
                'postalCode': homolog["CPF"],
                'uid': remove_zeros(homolog["matricula"]),
                'msSFU30Name': remove_zeros(homolog["matricula"]),
                'msSFU30NisDomain': 'campus',
                'uidNumber': get_msSFU30MaxUidNumber() + 1,
                'gidNumber': '10001',
                'loginShell': '/bin/bash',
                'unixHomeDirectory': f'/home/{remove_zeros(homolog["matricula"])}'
            }
    ):
        print(
            f'\nAluno ' + Bcolors.OKGREEN + f'{homolog["nome"]}' + Bcolors.ENDC + '\nMatricula'
            + Bcolors.OKGREEN + f' {homolog["matricula"]} ' + Bcolors.ENDC + 'adicionado no AD'
        )
        if conn.modify(ypserver, {"msSFU30MaxUidNumber": [(
                MODIFY_REPLACE, [f'{get_msSFU30MaxUidNumber() + 1}'])]}):
            print(
                f'\nO valor ' + Bcolors.OKGREEN + f'{get_msSFU30MaxUidNumber()}'
                + Bcolors.ENDC + ' foi modificado com sucesso!'
            )
```
No errors are returned. The user is created, but the unix attributes tab that should appear via RSAT view does not appear.

If I create the same user via RSAT, the unix attributes tab appears.

![Captura de tela de 2021-01-07 20-12-43](https://user-images.githubusercontent.com/29874574/103955362-dccc9680-5124-11eb-8053-b36ea986742d.png)

**Comment by zorn96** [comment]
hmm well this is interesting. so you're adding an entry to an AD domain right? (I'm guessing DC was domain controller in your previous comment)
I've never seen a "UNIX Attributes" tab in my AD experience

there's a few things you're doing here that I know have caused problems with AD in the past. I'm just going to throw out some bullet points. I can't be certain that any one is the cause of the problem offhand, but these are things to look at changing to see if it helps:

1. AD doesn't usually allow "cn" to be specified as an attribute, because it derives it from the DN. similarly, changing the CN in AD cannot be done through a normal modify and is instead done via a "move" operation (this differs from other servers like OpenLDAP)
2. uid is usually a multi-valued attribute in active directory, but here you're setting it as a single value. maybe AD accepts that and figures it out? or maybe python is doing the weird implicit conversion between strings and lists and messing your settings up. I haven't tried it myself

also some questions - what are msSFU30Name and msSFU30NisDomain? are those the fields you're expecting under "unix attributes"? how are they defined in your schema?
The reason I ask is because, from my AD experience, fields like uidNumber, unixHomeDirectory, and gidNumber all appear under "Attribute Editor" and not "Unix Attributes". Could you look there to see if the fields are set there properly, even if the tab doesn't appear?

basically, I suspect that the "Unix Attributes" tab might essentially be just a nicer way to filter what fields you see in order to view the unix fields that already exist under "Attribute Editor". But whether that tab appears or not might be something determined by your management tool, rather than something inherent to the record in AD, and so creating users via LDAP rather than management means the tab doesn't appear.
If this is the case, then you'll be able to see that the attributes are properly set if you go to "Attribute Editor". And the issue would be a gap with how the management tool interacts with records, rather than an issue in the ldap3 library

**Comment by zorn96** [comment]
actually it appears that this is the case. the "Unix Attributes" tab seems to be produced by the tool, not by the record, according to https://wiki.samba.org/index.php/Maintaining_Unix_Attributes_in_AD_using_ADUC


in newer versions of AD this tab is totally absent even with the tool 
https://wiki.samba.org/index.php/Installing_RSAT#Missing_Unix_Attributes_tab_in_ADUC_on_Windows_10_and_Windows_Server_2016
and instead you just need to look at the raw AD attributes. so my suspicion that the tool just cannot recognize the raw AD attributes to create the tab if you didn't use the tool to make the record seems more likely

here's a screenshot i took when i downloaded RSAT to use with windows 2019
<img width="414" alt="Screen Shot 2021-01-07 at 3 33 04 PM" src="https://user-images.githubusercontent.com/6414648/103957099-d1ff0b00-50fe-11eb-89cd-722359238b31.png">
<img width="1140" alt="Screen Shot 2021-01-07 at 3 39 43 PM" src="https://user-images.githubusercontent.com/6414648/103957103-d3c8ce80-50fe-11eb-89ae-44c921acff8f.png">


could you verify that the attributes appear as expected?

**Comment by empereira** [comment]
Nice... our RSAT is installed in win7 and unix attributes are visible.

I believe I did it. I modified the objectClass for this one `conn.add(DN, ['top', 'posixGroup', 'person', 'organizationalPerson', 'user']` and it worked. I compared with other users that were already correctly configured.

Now, the unix attributes tab showed up and I can verify the attributes that were configured by the script.
